airflow/jobs.py

```diff
@@ -403,39 +422,67 @@ def schedule_dag(self, dag):
                     dr.end_date = datetime.now()
             session.commit()
 
-            qry = session.query(func.max(DagRun.execution_date)).filter_by(
-                    dag_id = dag.dag_id).filter(
-                        or_(DagRun.external_trigger == False,
-                            # add % as a wildcard for the like query
-                            DagRun.run_id.like(DagRun.ID_PREFIX+'%')))
-            last_scheduled_run = qry.scalar()
+            # check last scheduled run
+            # get this dag's max execution date by
+            #  external trigger is false
+            #  and "like" Dag Run ID.ID
+            # bolke: why a like?
+            #qry = session.query(func.max(DagRun.execution_date)).filter_by(
+            #    dag_id=dag.dag_id).filter(
+            #    or_(DagRun.external_trigger is False,
+            #        # add % as a wildcard for the like query
+            #        DagRun.run_id.like(DagRun.ID_PREFIX+'%')
+            #        )
+            #)
+            #previous = qry.scalar()
+            previous = session.query(DagRun).filter_by(dag_id=dag.dag_id).filter(
+                or_(DagRun.external_trigger is False,
+                    DagRun.run_id.like(DagRun.ID_PREFIX+'%')
+                    )
+            ).order_by(DagRun.execution_date.desc()).first()
+
+            if previous:
+                self.logger.debug("Previous run {}".format(previous.execution_date))
+
+            # Determine next run date
             next_run_date = None
-            if dag.schedule_interval == '@once' and not last_scheduled_run:
+            if dag.schedule_interval == '@once' and not previous:
                 next_run_date = datetime.now()
-            elif not last_scheduled_run:
+            elif not previous:
                 # First run
-                TI = models.TaskInstance
-                latest_run = (
-                    session.query(func.max(TI.execution_date))
-                    .filter_by(dag_id=dag.dag_id)
-                    .scalar()
-                )
-                if latest_run:
-                    # Migrating from previous version
-                    # make the past 5 runs active
-                    next_run_date = dag.date_range(latest_run, -5)[0]
+                self.logger.debug("Scheduling {} for the first time".format(dag.dag_id))
+
+                # Set start_date based on tasks
+                task_start_dates = [t.start_date for t in dag.tasks]
+                if task_start_dates:
+                    # set next_run_date to start_date + interval unless start date
+                    # is on the interval
+                    min_task_start_date = min(task_start_dates)
+                    next_run_date = dag.following_schedule(min_task_start_date)
+                    next_run_date = self._normalize_schedule(dag,
```

Comment:
With `DAG.normalize_schedule` I suggested above you'd have a single simpler line of code here instead of 2 lines